### ISSUE TITLE
[RavenDB 3.5.10-patch-35295 and higher] Non-modified index raises InvalidOperationException: Cannot modify indexes

### ISSUE THREAD

**Issue by danielmarbach** [ravendb/ravendb#11004]
We are running into

> System.InvalidOperationException: Cannot modify indexes while indexing is in progress (already waited full minute). Try again later

even though the index hasn't changed. We have been able to track it down to the following commit

https://github.com/ravendb/ravendb/commit/b21ce6492e4deef8633641ec93e784040ef6a841#diff-248ebf22af1910601a7a3a507c9d0c8a

our index looks like

```
    public class MessagesViewIndex : AbstractIndexCreationTask<ProcessedMessage, MessagesViewIndex.SortAndFilterOptions>
    {
        public MessagesViewIndex()
        {
            Map = messages => from message in messages
                select new SortAndFilterOptions
                {
                    MessageId = (string)message.MessageMetadata["MessageId"],
                    MessageType = (string)message.MessageMetadata["MessageType"],
                    IsSystemMessage = (bool)message.MessageMetadata["IsSystemMessage"],
                    Status = (bool)message.MessageMetadata["IsRetried"] ? MessageStatus.ResolvedSuccessfully : MessageStatus.Successful,
                    TimeSent = (DateTime)message.MessageMetadata["TimeSent"],
                    ProcessedAt = message.ProcessedAt,
                    ReceivingEndpointName = ((EndpointDetails)message.MessageMetadata["ReceivingEndpoint"]).Name,
                    CriticalTime = (TimeSpan?)message.MessageMetadata["CriticalTime"],
                    ProcessingTime = (TimeSpan?)message.MessageMetadata["ProcessingTime"],
                    DeliveryTime = (TimeSpan?)message.MessageMetadata["DeliveryTime"],
                    Query = message.MessageMetadata.Select(_ => _.Value.ToString()).Union(new[] {string.Join(" ", message.Headers.Select(x => x.Value))}).ToArray(),
                    ConversationId = (string)message.MessageMetadata["ConversationId"]
                };

            Index(x => x.Query, FieldIndexing.Analyzed);

            Analyze(x => x.Query, typeof(StandardAnalyzer).AssemblyQualifiedName);

            DisableInMemoryIndexing = true;
        }
```

```
from message in docs.ProcessedMessages
select new {
	MessageId = ((string)message.MessageMetadata["MessageId"]),
	MessageType = ((string)message.MessageMetadata["MessageType"]),
	IsSystemMessage = ((bool)message.MessageMetadata["IsSystemMessage"]),
	Status = ((bool)message.MessageMetadata["IsRetried"]) ? 4 : 3,
	TimeSent = ((DateTime)message.MessageMetadata["TimeSent"]),
	ProcessedAt = message.ProcessedAt,
	ReceivingEndpointName = ((message.MessageMetadata["ReceivingEndpoint"])).Name,
	CriticalTime = ((TimeSpan?)message.MessageMetadata["CriticalTime"]),
	ProcessingTime = ((TimeSpan?)message.MessageMetadata["ProcessingTime"]),
	DeliveryTime = ((TimeSpan?)message.MessageMetadata["DeliveryTime"]),
	Query = DynamicEnumerable.ToArray(DynamicEnumerable.Union(message.MessageMetadata.Select(_ => _.Value.ToString()), new string[] {
		String.Join(" ", message.Headers.Select(x => x.Value))
	})),
	ConversationId = ((string)message.MessageMetadata["ConversationId"])
}
```

and yes it is always this index and only that one (the only one that contains `ToArray()`)

![image](https://user-images.githubusercontent.com/174258/90170704-b6d2de00-dda0-11ea-9d9d-5722034d801a.png)

as soon as we downgrade to RavenDB 3.5.10-patch-35294 the problem goes away. We are going to need to stop updating to the latest updates of the 3.5 versions for now until this issue is resolved.

## Exception

```
2020-08-12 07:43:03.5509|14|Warn|Raven.Client.Connection.Async.AsyncServerClient|Was unable to fetch topology from primary node http://localhost also there is no cached topology
2020-08-12 07:44:03.8485|37|Warn|Raven.Database.DocumentDatabase|Could not create index batch
System.InvalidOperationException: Cannot modify indexes while indexing is in progress (already waited full minute). Try again later
   at Raven.Database.Storage.IndexDefinitionStorage.TryRemoveIndexContext()
   at Raven.Database.Actions.IndexActions.DeleteIndex(IndexDefinition instance, Boolean removeByNameMapping, Boolean clearErrors, Boolean removeIndexReplaceDocument, Boolean isSideBySideReplacement, Nullable`1 deletedIndexVersion)
   at Raven.Database.Actions.IndexActions.DeleteIndex(String name, Nullable`1 deletedIndexVersion)
   at Raven.Database.Actions.IndexActions.PutIndexInternal(String name, IndexDefinition definition, Boolean disableIndexBeforePut, Boolean isUpdateBySideSide, Nullable`1 creationOptions, Boolean isReplication)
   at Raven.Database.Actions.IndexActions.PutIndexes(IndexToAdd[] indexesToAdd, Boolean isReplication)
2020-08-12 07:45:03.9962|13|Error|Raven.Database.Server.Controllers.RavenBaseApiController|Cannot create index.
System.InvalidOperationException: Cannot modify indexes while indexing is in progress (already waited full minute). Try again later
   at Raven.Database.Storage.IndexDefinitionStorage.TryRemoveIndexContext()
   at Raven.Database.Actions.IndexActions.DeleteIndex(IndexDefinition instance, Boolean removeByNameMapping, Boolean clearErrors, Boolean removeIndexReplaceDocument, Boolean isSideBySideReplacement, Nullable`1 deletedIndexVersion)
   at Raven.Database.Actions.IndexActions.DeleteIndex(String name, Nullable`1 deletedIndexVersion)
   at Raven.Database.Actions.IndexActions.PutIndexInternal(String name, IndexDefinition definition, Boolean disableIndexBeforePut, Boolean isUpdateBySideSide, Nullable`1 creationOptions, Boolean isReplication)
   at Raven.Database.Actions.IndexActions.PutIndex(String name, IndexDefinition definition, Boolean isReplication)
   at Raven.Database.Server.Controllers.IndexController.<IndexPut>d__5.MoveNext()
2020-08-12 07:45:04.0682|29|Error|Raven.Database.Impl.BackgroundTaskExecuter.RavenThreadPool|Error occured while executing RavenThreadPool task ; Database name:  ; Task queued at: 8/12/2020 7:45:04 AM ; Task Description: Mapping index Raven/DocumentsByEntityName from etag 01000000-0000-0003-0000-000000009F64 to etag 01000000-0000-0003-0000-00000000AF64 range 1024 to 2047 of 4096
System.OperationCanceledException: The operation was canceled.
   at System.Threading.CancellationToken.ThrowOperationCanceledException()
   at Raven.Database.Indexing.RobustEnumerator.<RobustEnumeration>d__7.MoveNext()
   at Raven.Database.Indexing.RobustEnumerator.<RobustEnumeration>d__8.MoveNext()
   at Raven.Database.Indexing.SimpleIndex.<>c__DisplayClass7_1.<IndexDocuments>b__8(IEnumerator`1 partition)
   at Raven.Database.Impl.BackgroundTaskExecuter.RavenThreadPool.<>c__DisplayClass38_0`1.<ExecuteBatch>b__0()
   at Raven.Database.Impl.BackgroundTaskExecuter.RavenThreadPool.RunThreadTask(ThreadTask threadTask)
2020-08-12 07:45:04.0682|23|Error|Raven.Database.Impl.BackgroundTaskExecuter.RavenThreadPool|Error occured while executing RavenThreadPool task ; Database name:  ; Task queued at: 8/12/2020 7:45:04 AM ; Task Description: Mapping index Raven/DocumentsByEntityName from etag 01000000-0000-0003-0000-000000009F64 to etag 01000000-0000-0003-0000-00000000AF64 range 0 to 1023 of 4096
System.OperationCanceledException: The operation was canceled.
   at System.Threading.CancellationToken.ThrowOperationCanceledException()
   at Raven.Database.Indexing.RobustEnumerator.<RobustEnumeration>d__7.MoveNext()
   at Raven.Database.Indexing.RobustEnumerator.<RobustEnumeration>d__8.MoveNext()
   at Raven.Database.Indexing.SimpleIndex.<>c__DisplayClass7_1.<IndexDocuments>b__8(IEnumerator`1 partition)
   at Raven.Database.Impl.BackgroundTaskExecuter.RavenThreadPool.<>c__DisplayClass38_0`1.<ExecuteBatch>b__0()
   at Raven.Database.Impl.BackgroundTaskExecuter.RavenThreadPool.RunThreadTask(ThreadTask threadTask)
2020-08-12 07:45:04.0682|38|Error|Raven.Database.Impl.BackgroundTaskExecuter.RavenThreadPool|Error occured while executing RavenThreadPool task ; Database name:  ; Task queued at: 08/12/2020 07:45:04 ; Task Description: Mapping index Raven/DocumentsByEntityName from etag 01000000-0000-0003-0000-000000009F64 to etag 01000000-0000-0003-0000-00000000AF64 range 2048 to 3071 of 4096
System.OperationCanceledException: The operation was canceled.
   at System.Threading.CancellationToken.ThrowOperationCanceledException()
   at Raven.Database.Indexing.RobustEnumerator.<RobustEnumeration>d__7.MoveNext()
   at Raven.Database.Indexing.RobustEnumerator.<RobustEnumeration>d__8.MoveNext()
   at Raven.Database.Indexing.SimpleIndex.<>c__DisplayClass7_1.<IndexDocuments>b__8(IEnumerator`1 partition)
   at Raven.Database.Impl.BackgroundTaskExecuter.RavenThreadPool.<>c__DisplayClass38_0`1.<ExecuteBatch>b__0()
   at Raven.Database.Impl.BackgroundTaskExecuter.RavenThreadPool.RunThreadTask(ThreadTask threadTask)
2020-08-12 07:45:04.0682|25|Error|Raven.Database.Impl.BackgroundTaskExecuter.RavenThreadPool|Error occured while executing RavenThreadPool task ; Database name:  ; Task queued at: 8/12/2020 7:45:04 AM ; Task Description: Mapping index Raven/DocumentsByEntityName from etag 01000000-0000-0003-0000-000000009F64 to etag 01000000-0000-0003-0000-00000000AF64 range 3072 to 4095 of 4096
System.OperationCanceledException: The operation was canceled.
   at System.Threading.CancellationToken.ThrowOperationCanceledException()
   at Raven.Database.Indexing.RobustEnumerator.<RobustEnumeration>d__7.MoveNext()
   at Raven.Database.Indexing.RobustEnumerator.<RobustEnumeration>d__8.MoveNext()
   at Raven.Database.Indexing.SimpleIndex.<>c__DisplayClass7_1.<IndexDocuments>b__8(IEnumerator`1 partition)
   at Raven.Database.Impl.BackgroundTaskExecuter.RavenThreadPool.<>c__DisplayClass38_0`1.<ExecuteBatch>b__0()
   at Raven.Database.Impl.BackgroundTaskExecuter.RavenThreadPool.RunThreadTask(ThreadTask threadTask)
2020-08-12 07:45:04.0832|1|Warn|Raven.Database.Prefetching.PrefetchingBehavior|WASTE: Discarding future work item without using it, to reduce memory usage
2020-08-12 07:45:04.0832|38|Error|Raven.Database.Impl.BackgroundTaskExecuter.RavenThreadPool|Error occured while executing RavenThreadPool task ; Database name:  ; Task queued at: 8/12/2020 7:45:04 AM ; Task Description: Performing indexing on index batches for a total of 2 indexes range 1 to 1 of 2
System.AggregateException: One or more errors occurred. ---> System.AggregateException: One or more errors occurred. ---> System.Threading.Tasks.TaskCanceledException: A task was canceled.
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task`1.GetResultCore(Boolean waitCompletionNotification)
   at Raven.Database.Prefetching.PrefetchingBehavior.<>c__DisplayClass97_0.<AddFutureBatch>b__2(Task`1 t)
   at System.Threading.Tasks.ContinuationResultTaskFromResultTask`2.InnerInvoke()
   at System.Threading.Tasks.Task.Execute()
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task`1.GetResultCore(Boolean waitCompletionNotification)
   at Raven.Database.Prefetching.PrefetchingBehavior.<>c__DisplayClass113_0.<UpdateAutoThrottler>b__0(FutureIndexBatch x)
   at System.Linq.Enumerable.WhereSelectEnumerableIterator`2.MoveNext()
   at System.Linq.Enumerable.Sum(IEnumerable`1 source)
   at Raven.Database.Prefetching.PrefetchingBehavior.UpdateAutoThrottler(List`1 jsonDocs, TimeSpan indexingDuration, Boolean skipIncreasingBatchSize)
   at Raven.Database.Indexing.IndexingExecuter.ReleasePrefetchersAndUpdateStatistics(IndexingGroup indexingGroup, TimeSpan elapsedTimeSpan)
   at Raven.Database.Indexing.IndexingExecuter.<>c__DisplayClass19_1.<ExecuteIndexingWork>b__5(IndexingGroup x)
   at Raven.Database.Indexing.IndexingExecuter.IndexingGroup.SignalIndexingComplete()
   at Raven.Database.Indexing.IndexingExecuter.IndexingBatchForIndex.SignalIndexingComplete()
   at Raven.Database.Indexing.IndexingExecuter.HandleIndexingFor(IndexingBatchForIndex batchForIndex, Etag lastEtag, DateTime lastModified, CancellationToken token)
   at Raven.Database.Indexing.IndexingExecuter.<PerformIndexingOnIndexBatches>b__25_0(IndexingBatchOperation indexBatchOperation)
   at Raven.Database.Impl.BackgroundTaskExecuter.RavenThreadPool.<>c__DisplayClass41_1`1.<ExecuteBatch>b__0()
   at Raven.Database.Impl.BackgroundTaskExecuter.RavenThreadPool.RunThreadTask(ThreadTask threadTask)
---> (Inner Exception #0) System.AggregateException: One or more errors occurred. ---> System.Threading.Tasks.TaskCanceledException: A task was canceled.
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task`1.GetResultCore(Boolean waitCompletionNotification)
   at Raven.Database.Prefetching.PrefetchingBehavior.<>c__DisplayClass97_0.<AddFutureBatch>b__2(Task`1 t)
   at System.Threading.Tasks.ContinuationResultTaskFromResultTask`2.InnerInvoke()
   at System.Threading.Tasks.Task.Execute()
---> (Inner Exception #0) System.Threading.Tasks.TaskCanceledException: A task was canceled.<---
<---
2020-08-12 07:59:01.8993|38|Error|Raven.Database.DocumentDatabase|Index '6-(Unknown Name)' removed because it is not found or invalid
2020-08-12 07:59:03.0802|16|Warn|Raven.Client.Connection.Async.AsyncServerClient|Was unable to fetch topology from primary node http://localhost also there is no cached topology
2020-08-12 08:00:03.3755|16|Warn|Raven.Database.DocumentDatabase|Could not create index batch
System.InvalidOperationException: Cannot modify indexes while indexing is in progress (already waited full minute). Try again later
   at Raven.Database.Storage.IndexDefinitionStorage.TryRemoveIndexContext()
   at Raven.Database.Actions.IndexActions.DeleteIndex(IndexDefinition instance, Boolean removeByNameMapping, Boolean clearErrors, Boolean removeIndexReplaceDocument, Boolean isSideBySideReplacement, Nullable`1 deletedIndexVersion)
   at Raven.Database.Actions.IndexActions.DeleteIndex(String name, Nullable`1 deletedIndexVersion)
   at Raven.Database.Actions.IndexActions.PutIndexInternal(String name, IndexDefinition definition, Boolean disableIndexBeforePut, Boolean isUpdateBySideSide, Nullable`1 creationOptions, Boolean isReplication)
   at Raven.Database.Actions.IndexActions.PutIndexes(IndexToAdd[] indexesToAdd, Boolean isReplication)
2020-08-12 08:01:03.5204|18|Error|Raven.Database.Server.Controllers.RavenBaseApiController|Cannot create index.
System.InvalidOperationException: Cannot modify indexes while indexing is in progress (already waited full minute). Try again later
   at Raven.Database.Storage.IndexDefinitionStorage.TryRemoveIndexContext()
   at Raven.Database.Actions.IndexActions.DeleteIndex(IndexDefinition instance, Boolean removeByNameMapping, Boolean clearErrors, Boolean removeIndexReplaceDocument, Boolean isSideBySideReplacement, Nullable`1 deletedIndexVersion)
   at Raven.Database.Actions.IndexActions.DeleteIndex(String name, Nullable`1 deletedIndexVersion)
   at Raven.Database.Actions.IndexActions.PutIndexInternal(String name, IndexDefinition definition, Boolean disableIndexBeforePut, Boolean isUpdateBySideSide, Nullable`1 creationOptions, Boolean isReplication)
   at Raven.Database.Actions.IndexActions.PutIndex(String name, IndexDefinition definition, Boolean isReplication)
   at Raven.Database.Server.Controllers.IndexController.<IndexPut>d__5.MoveNext()
```

**Comment by ayende** [comment]
Do you have your index _always_ busy? This commit is related to improving how we can handle `ToArray()` is some cases.
The index should automatically update itself when it has time.

A workaround may be to change the index definition to use the old definition as a string.

**Comment by danielmarbach** [comment]
@ayende They might be. Plus as it seems this change causes our MessageViewIndex to change and therefore reindexing will happen. Some of our customers have hundreds of GB of data. 

Are you saying because of afaik a breaking change that causes reindexing for all customers that use `ToArray` you are suggesting us to fall back to use string-based index definitions?

**Comment by ayende** [comment]
The quick fix is to use the string version, yes. 
Note that 3.5 is on low maintenance mode, so the next bug fix release is not scheduled any time soon.

This shouldn't actually cause you an issue, aside from the re-indexing work. Note that it is actually fairly common for the index text output to change with client versions, as long as the results are equivalent, there is no compatability concern from our perspective.

**Comment by grisha-kotler** [comment]
@danielmarbach is it .NET Framework or .NET Core?

**Comment by danielmarbach** [comment]
.Net Framework. I'm aware that this version is on low maintenance mode and we are in preparation to update to later major versions. Still causing several hundreds of customers to go through reindexing of several hundred gigs of data seems a bit excessive especially because these changes were sneaked in in a patch. For now we are stuck unless we switch to string indexes.

As a feed back it wouldbe great if changes like that are clearly marked in the changelog as potentially breaking or let's say causing to reindexing. We read the changes carefully before we bump the dependency

**Comment by grisha-kotler** [comment]
https://github.com/ravendb/ravendb/pull/11051

**Comment by grisha-kotler** [comment]
We released a new patch build, 3.5.10-patch-35301
http://hibernatingrhinos.com/builds?product=ravendb&version=3.5&channel=stable

**Comment by danielmarbach** [comment]
Thank you!
